### PR TITLE
Feat/token permissions pending state

### PR DESCRIPTION
**Pull request by osmaczko**
### What does the PR do

closes: https://github.com/status-im/status-desktop/issues/11852

status-go PR: https://github.com/status-im/status-go/pull/3942

### Affected areas

- community token permissions management

### Video of functionality

https://github.com/status-im/status-desktop/assets/33099791/7364d09e-f583-430a-98fd-1290fc3df5f7

### CODE REVIEW

**Comment by osmaczko** [comment]
I've just spotted an issue. Everything works fine as long as the control node is online. However, if the control node goes offline and an admin makes changes to the community, these changes are not applied when the control node comes back online. They will only be applied when the control node makes some changes. I am investigating whether this is due to the recent changes or if it was like this before.

CC: @jrainville

**Comment by osmaczko** [comment]
> I've just spotted an issue. Everything works fine as long as the control node is online. However, if the control node goes offline and an admin makes changes to the community, these changes are not applied when the control node comes back online. They will only be applied when the control node makes some changes. I am investigating whether this is due to the recent changes or if it was like this before.
> 
> CC: @jrainville

Added logs, when control node becomes online it rejects the events, causing an endless reject event <-> resend event loop. The loop can be quit only when control node propagates new changes.
```
HandleCommunityEventsMessage, err: clock for admin event message is outdated
HandleCommunityEventsMessage, err: clock for admin event message is outdated
HandleCommunityEventsMessage, err: clock for admin event message is outdated
HandleCommunityEventsMessage, err: clock for admin event message is outdated
HandleCommunityEventsMessage, err: clock for admin event message is outdated
HandleCommunityEventsMessage, err: clock for admin event message is outdated
HandleCommunityEventsMessage, err: clock for admin event message is outdated
HandleCommunityEventsMessage, err: clock for admin event message is outdated
HandleCommunityEventsMessage, err: clock for admin event message is outdated
HandleCommunityEventsMessage, err: clock for admin event message is outdated
HandleCommunityEventsMessage, err: clock for admin event message is outdated
HandleCommunityEventsMessage, err: clock for admin event message is outdated
```

**Comment by jrainville** [comment]
Nice find @osmaczko .

I did think I spotted that when I worked on the pending join requests, but I didn't understand when and what was happening. 

If you want, you can open an issue and we can work on it in another PR

**Comment by osmaczko** [comment]
> I've just spotted an issue. Everything works fine as long as the control node is online. However, if the control node goes offline and an admin makes changes to the community, these changes are not applied when the control node comes back online. They will only be applied when the control node makes some changes. I am investigating whether this is due to the recent changes or if it was like this before.
> 
> CC: @jrainville

Found the root cause. It is working now.

https://github.com/status-im/status-desktop/assets/33099791/87aef20b-d386-4b89-9b9c-0c8d61df9ace

**Review by jrainville — APPROVED** [review]
Good job!

File: src/app_service/service/community/dto/community.nim [hunk 1]
```diff
@@ -283,6 +290,11 @@ proc toCommunityTokenPermissionDto*(jsonObj: JsonNode): CommunityTokenPermission
   if (tokenPermissionTypeInt >= ord(low(TokenPermissionType)) or tokenPermissionTypeInt <= ord(high(TokenPermissionType))):
       result.`type` = TokenPermissionType(tokenPermissionTypeInt)
 
+  var tokenPermissionStateInt: int
+  discard jsonObj.getProp("state", tokenPermissionStateInt)
+  if (tokenPermissionStateInt >= ord(low(TokenPermissionState)) or tokenPermissionStateInt <= ord(high(TokenPermissionState))):
```
Comment (caybro) [inline src/app_service/service/community/dto/community.nim:295]:
```suggestion
  if (tokenPermissionStateInt >= ord(low(TokenPermissionState)) and tokenPermissionStateInt <= ord(high(TokenPermissionState))):
```

**Comment by osmaczko** [comment]
> One thing noted is that it's generating a lot of toasts. Perhaps some after the permission is updated locally and then again after the control node updates? 

I agree, that's annoying. The toasts seems to be duplicated because indeed the permission is changed that many times, that is: updated -> pending -> approved by control node - all these states are toasted as "pending".

> Will mark as an issue but this has been open a long time now so we could split it as well.

Thanks, I also think it should be addressed it in a separate issue. I've created one: https://github.com/status-im/status-desktop/issues/12184

File: src/app_service/service/community/dto/community.nim (hunk 1 above)
Comment (osmaczko) [reply src/app_service/service/community/dto/community.nim:295]: Nice catch, I blindly copied it from the line above :sweat_smile:

Comment (osmaczko) [reply src/app_service/service/community/dto/community.nim]: Fixed

**Review by MishkaRogachev — APPROVED** [review]

**Pull request merged**